src/main/scala/com/fulcrumgenomics/vcf/filtration/ReadEndSomaticVariantFilter.scala

```diff
@@ -25,48 +25,49 @@
-  *
-  * @param distance The distance from the end of the template that defines the region where this
-  *                 artifact may occur
-  * @param pValueThreshold the pvalue at and below which mutations are filtered as being likely
-  *                        caused by the artifact
-  */
-class EndRepairArtifactLikelihoodFilter(val distance: Int = 2,
-                                        pValueThreshold: Option[Double] = None) extends SomaticVariantFilter with LazyLogging {
 
-  /** The INFO header line needed by this filter. */
-  val Info = VcfInfoHeader("ERAP", VcfCount.Fixed(1), VcfFieldType.Float, "P-Value for the End-Repair Artifact test.")
+/** Trait for classes that apply filters to somatic variants calls at read ends. */
+trait ReadEndSomaticVariantFilter extends SomaticVariantFilter with LazyLogging {
 
-  /** The FILTER header line needed by this filter. */
-  val Filter = VcfFilterHeader("EndRepairArtifact", "Variant is likely an artifact caused by end-repair and A-base addition.")
+  /** The VCF header line that describes the INFO key that the filter may reference. */
+  val Info: VcfInfoHeader
 
-  override val VcfInfoLines: Iterable[VcfInfoHeader]     = Seq(Info)
-  override val VcfFilterLines: Iterable[VcfFilterHeader] = Seq(Filter)
+  /** The VCF header line that describes the FILTER key that the filter may reference. */
+  val Filter: VcfFilterHeader
 
-  /** Only applies to het SNVs where the alt allele is A or T. */
-  override def appliesTo(gt: Genotype): Boolean = {
-    gt.isHet && !gt.isHetNonRef && gt.calls.forall(_.value.length == 1) &&
-      gt.calls.iterator.filter(_ != gt.alleles.ref).exists(a => a.value.equalsIgnoreCase("A") || a.value.equalsIgnoreCase("T"))
+  /** The distance from the end of the template that defines the region where the artifact may occur */
+  val distance: Int
+
+  /** The pvalue at or below which mutations are filtered as being likely caused by the artifact */
+  val pValueThreshold: Option[Double] = None
+
+  /** Function for testing whether a base in the pileup (either ref or alt) has properties that are
+    * congruent with it having formed due to the artifact-generating process germane to the filter.
+    * Depending on the filter, this may be as simple as testing whether the base is positioned within
+    * a threshold distance from the read end. Other filters employ more elaborate tests that consider
+    * multiple criteria, such as the presence of certain allele pairs at specific read ends.
+    */
+  def isArtifactCongruent(refAllele: Byte, altAllele: Byte, entry : BaseEntry, pileupPosition: Int): Boolean
+
+  /** Applies a single filter if a) a threshold was provided at construction, b) the computed
+    * pvalue is present in the <annotations>, and c) the pvalue <= <pValueThreshold>.
+    */
+  def filters(annotations: Map[String, Any]): Iterable[String] = {
+    (this.pValueThreshold, annotations.get(this.VcfInfoLines.head.id).map(_.asInstanceOf[Double])) match {
```

Comment:
Why just the head (first record) of `VcfInfoLines`?

Reply:
Any way to avoid the `asInstanceOf` cast? Which may fail at runtime?

Reply:
Thanks to a really valuable peer programming session with @mjhipp, both of these are now addressed and I have a better understanding of some Scala fundamentals, specifically "`Map` vs. `flatMap`" and "why `.get` is sometimes dangerous."